File: java/all-your-base/src/main/java/BaseConverter.java
public class BaseConverter {
    private int decimalValue = 0;

    BaseConverter(int fromBase, int[] digits) {
        if (fromBase <= 1)
            throw new IllegalArgumentException("Bases must be at least 2.");

        for (int i = 0; i < digits.length; i++) {
            if (digits[i] < 0)
                throw new IllegalArgumentException("Digits may not be negative.");
            if (digits[i] >= fromBase)
                throw new IllegalArgumentException("All digits must be strictly less than the base.");

            decimalValue = decimalValue * fromBase + digits[i];
        }
    }

    int[] convertToBase(int toBase) {
        if (toBase <= 1)
            throw new IllegalArgumentException("Bases must be at least 2.");

        if (decimalValue == 0) return new int[]{0};

        int toDigitsLength = 1 + (int)Math.floor(Math.log10(decimalValue) / Math.log10(toBase));
        int[] result = new int[toDigitsLength];

        for (int i = toDigitsLength - 1; decimalValue > 0; i--) {
            result[i] = decimalValue % toBase;
            decimalValue /= toBase;
        }
        return result;
    }
}

File: kotlin/all-your-base/src/main/kotlin/BaseConverter.kt
class BaseConverter(val fromBase: Int, val digits: IntArray) {
    init {
        require(fromBase > 1) { "Bases must be at least 2." }
        require(digits.isNotEmpty()) { "You must supply at least one digit." }
        require(!(digits.size > 1 && digits[0] == 0)) { "Digits may not contain leading zeros." }
        require(digits.all { it >= 0 }) { "Digits may not be negative." }
        require(digits.all { it < fromBase }) { "All digits must be strictly less than the base." }
    }

    private val decimal = digits.reduce { sum, d -> sum * fromBase + d }

    fun convertToBase(newBase: Int): IntArray {
        require(newBase > 1) { "Bases must be at least 2." }

        if (decimal < newBase) return intArrayOf(decimal)

        var d = decimal
        val newDigits = mutableListOf<Int>()
        while (d > 0) {
            newDigits.add(0, d % newBase)
            d /= newBase
        }
        return newDigits.toIntArray()
    }
}
